### PR TITLE
add anatomy of variables

### PR DESCRIPTION
**Pull request by mradeybee**

### CODE REVIEW

File: anatomy_of_variables/Anatomy of Variables.md [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+end
+
+count #=> 0 1 2 3 4 5 6 7 8 9 10
+```
+
+This is a ruby method named `counter`. A local variable `count` was assigned in this method and it was used to print `0..10`. However if you call count outside this method, it would throw a `method or variable not defined` error.
+
+Here is the JavaScript version of the above code:
+
+```JavaScript
```
Comment (ghost) [inline anatomy_of_variables/Anatomy of Variables.md:46]: Is there a specific reason why you are writing the same function in javascript. Does it really make a difference if we have it here?

File: anatomy_of_variables/Anatomy of Variables.md [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    end
+end
+
+counter(0); #=>0 1 2 3 4 5 6 7 8 9 10
+```
+
+Here is the JavaScript version of the code above.
+
+```JavaScript
+function counter(count){
```
Comment (ghost) [inline anatomy_of_variables/Anatomy of Variables.md:74]: Same with above

File: anatomy_of_variables/Anatomy of Variables.md [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+```
+
+**Global Variables**
+
+These are variables that can be used by two or more methods or function. These variables are assigned outside the methods or functions and they can be manipulated within methods or functions.
+
+The example below shows how global variables are assigned and used in Ruby and JavaScript.
+
+```ruby
+@count = 0
```
Comment (ghost) [inline anatomy_of_variables/Anatomy of Variables.md:92]: Global variables in ruby starts with $. So we can have `$count`. @count is a different form of global variables called `instance variables`.

File: anatomy_of_variables/Anatomy of Variables.md [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+The example below shows an instance variable been used in an instance method.
+
+```ruby
+class User
+    def initialize
+        @name = "Adebayo"
+    end
+
+    def greet
+        puts "Hello #{@name}"
```
Comment (ghost) [inline anatomy_of_variables/Anatomy of Variables.md:142]: I am not sure using the `puts` here is necessary. What are your thoughts?

File: anatomy_of_variables/Anatomy of Variables.md [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+        puts @count
+        @count += 1
+    end
+end
+
+counter #=>0 1 2 3 4 5 6 7 8 9 10
+```
+
+
+```JavaScript
```
Comment (ghost) [inline anatomy_of_variables/Anatomy of Variables.md:105]: Same with above.

File: anatomy_of_variables/Anatomy of Variables.md [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+end
+
+User.new.greet #=> Hello Adebayo
+```
+
+The above example has instance variable `@name` which holds a name `Adebayo`. The instance method `greet` adds `Hello` to the value of `@name`.  This tells us that instance variables act as global variables **within** a class.  
+
+**Class Variables**
+
+Class variables work exactly like instance methods. These varibales are denoted with a double `@` sign, i.e `@@`. These variables are  **ONLY** used in **Class Methods** within a class. Just in case, you don't know what a class method is in ruby, let's take a quick look.
```
Comment (ghost) [inline anatomy_of_variables/Anatomy of Variables.md:153]: Class variables can be used in `instance methods`. Infact, they can be used anywhere within the class or it's children. They are the closest to the regular `$variable` global variable.

File: anatomy_of_variables/Anatomy of Variables.md [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+
+The keyword `self` makes a class method quite diferent from an instance method. A class method is also defined by the `def` keyword, then the keyword `self` is prpended to the method name. For example, if we were to rewrite the above example as a class method, the method `greet` will be written as `self.greet`. Class methods, however, can be called directly on the class they belong to. The example below shows a class variable been used in a class method.
```
Comment (ghost) [inline anatomy_of_variables/Anatomy of Variables.md:155]: prpended => prepended. 
Can you rephrase the use of self in this context. `self` can mean different things in different context. Check out [this blog post](https://www.jimmycuadra.com/posts/self-in-ruby/) to learn more about self. I will also be writing on self in the nearest future.

File: anatomy_of_variables/Anatomy of Variables.md [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+         puts "Hello #{@@name}"
+    end
+end
+
+User.name  #=> "Hello Adebayo"
+```
+
+**Class Instance Variables**
+
+These variables are a combination of the two we have discussed earlier. They are used as global variables but they are assigned as local variables. This means they are assigned in a class method like instance variables but they can be used in any class methods within the same class.
```
Comment (ghost) [inline anatomy_of_variables/Anatomy of Variables.md:171]: What do you mean by `They are used as global variables but they are assigned as local variables`? Can you clarify?

File: anatomy_of_variables/Anatomy of Variables.md [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+end
+
+Person.username("Adebayo") #=> "Adebayo"
+Person.greet #=> "Hello Adebayo"
+
+```
+
+In the snippet above, `name` method was used to get the person's name and it was assigned to `@name` variable. This variable, however, can be used in any other class method within that class.
+
+One important thing to know is that this kind of variables will not be available in classes that inherit from the class in which they are assigned.
```
Comment (ghost) [inline anatomy_of_variables/Anatomy of Variables.md:193]: Your examples didn't prove that they are not available.
Just like you need to call `username` method on Person class before invoking `greet` method, you have to call `username` method in Property class before you invoke `owner` method.

A good example will be

``` ruby
 class Person
   @name = "Adebayo"
   def self.greet
       "Hello #{@name}"
    end
 end

Person.greet #=> "Hello Adebayo"

class Student < Person; end
Student.greet #=> "Hello "

class Student < Person
  @name = "Penmuse"
end

Student.greet #=> "Hello Penmuse"
```

File: anatomy_of_variables/Anatomy of Variables.md [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+The major differences between these variables wrap around the scope at which they are to be used and the syntax for assigning them.
+
+* The diference between an instance variable and a class instance variable is the scope at which they are been used.
+
+    > Instance Variables are scoped to Instance Methods while Class Instance Variable are scoped to Class Methods.
+
+* The diference between a Class Instance Variable and a Class Variable is the syntax of assinging them.
+
+    > Class instance Variables are assigned with a single `@` symbol while Class Variables are assigned with double `@` symbol, i.e `@@`
```
Comment (ghost) [inline anatomy_of_variables/Anatomy of Variables.md:229]: I feel it always makes sense to have a closing remark like conclusion or final thoughts. What do you think of having such?

File: anatomy_of_variables/Anatomy of Variables.md (hunk 1 above)
Comment (mradeybee) [inline anatomy_of_variables/Anatomy of Variables.md:46]: There is no actual reason for writing the same code in `JS`. I thought of removing it too, since this is a ruby tutorial.

File: anatomy_of_variables/Anatomy of Variables.md (hunk 4 above)
Comment (mradeybee) [inline anatomy_of_variables/Anatomy of Variables.md:142]: Yeah, it's really not necessary

File: anatomy_of_variables/Anatomy of Variables.md (hunk 6 above)
Comment (mradeybee) [inline anatomy_of_variables/Anatomy of Variables.md:153]: You are right sir, I got it mixed up with class instance variables.

File: anatomy_of_variables/Anatomy of Variables.md (hunk 8 above)
Comment (mradeybee) [inline anatomy_of_variables/Anatomy of Variables.md:171]: So I think the statement is wrong. I will try and rephrase.

File: anatomy_of_variables/Anatomy of Variables.md (hunk 10 above)
Comment (mradeybee) [inline anatomy_of_variables/Anatomy of Variables.md:229]: Sounds nice.
